Invalidate :nth-child() selectors correctly when not in subject position

heycam opened the pull request:
#### 0c0bfc05ec5c27a7bd2176002fd581a4f3157f7e
<pre>
Invalidate :nth-child() selectors correctly when not in subject position
<a href="https://bugs.webkit.org/show_bug.cgi?id=253941">https://bugs.webkit.org/show_bug.cgi?id=253941</a>
rdar://106740353

Reviewed by Antti Koivisto.

We use MatchResult::AnySibling when an element that matches the selector
inside an :nth-child() pseudo-class changes. But when we process that
invalidation, it only invalidates the element itself. This means that
AnySibling is only correct when the :nth-child() is in the subject
position.

Introduce MatchResult::ParentAnySibling and AncestorAnySibling to
represent invalidations required for :nth-child() pseudo-classes
followed by child or descendant combinators.

Canonical link: <a href="https://commits.webkit.org/267600@main">https://commits.webkit.org/267600@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/511b27d3a22328a9d54357ff1223e4b4502f20d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18914 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16026 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17320 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20727 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17595 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18225 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17680 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19730 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15557 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22244 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15919 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15724 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20061 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16314 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15468 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4087 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16145 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/ClassChangeInvalidation.cpp:
```diff
@@ -115,15 +115,27 @@ void ClassChangeInvalidation::computeInvalidation(const SpaceSplitString& oldCla
             return true;
-        return type == ClassChangeType::Add ? isNegation == IsNegation::No : isNegation == IsNegation::Yes;
+        default:
+            return type == ClassChangeType::Add ? isNegation == IsNegation::No : isNegation == IsNegation::Yes;
```
It might be better to remove the default and list all values explicitly here so next person who adds a value has to consider this.

heycam committed:
Invalidate :nth-child() selectors correctly when not in subject position
https://bugs.webkit.org/show_bug.cgi?id=253941
rdar://106740353

Reviewed by Antti Koivisto.

We use MatchResult::AnySibling when an element that matches the selector
inside an :nth-child() pseudo-class changes. But when we process that
invalidation, it only invalidates the element itself. This means that
AnySibling is only correct when the :nth-child() is in the subject
position.

Introduce MatchResult::ParentAnySibling and AncestorAnySibling to
represent invalidations required for :nth-child() pseudo-classes
followed by child or descendant combinators.

Canonical link: https://commits.webkit.org/267600@main
